internal/lsp/server.go

```diff
@@ -204,6 +219,74 @@ func (l *LanguageServer) StartHoverWorker(ctx context.Context) {
 	}
 }
 
+func (l *LanguageServer) StartCommandWorker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case params := <-l.commandRequest:
+			switch params.Command {
+			case "regal.fmt":
+				if len(params.Arguments) == 0 {
+					l.logError(fmt.Errorf("expected at least one argument in command %v", params.Arguments))
+
+					break
+				}
+
+				target, ok := params.Arguments[0].(string)
+				if !ok {
+					l.logError(fmt.Errorf("expected argument to be a string in command %v", params.Command))
+
+					break
+				}
+
+				oldContent, ok := l.cache.GetFileContents(target)
+				if !ok {
+					l.logError(fmt.Errorf("could not get file contents for uri %q", target))
+
+					break
+				}
+
+				newContent, err := Format(uri.ToPath(l.clientIdentifier, target), oldContent, format.Opts{})
+				if err != nil {
+					l.logError(fmt.Errorf("failed to format file: %w", err))
+
+					break
+				}
+
+				edits := ComputeEdits(oldContent, newContent)
+
+				editParams := ApplyWorkspaceEditParams{
+					Label: "Format using opa fmt",
+					Edit: WorkspaceEdit{
+						DocumentChanges: []TextDocumentEdit{
+							{
+								TextDocument: TextDocumentIdentifier{URI: target},
+								Edits:        edits,
+							},
+						},
+					},
+				}
+
+				l.logOutboundMessage(methodWorkspaceApplyEdit, editParams)
+
+				// note, here conn.Call is used as the workspace/applyEdit message is a request, not a notification
+				// as per the spec. In order to be 'routed' to the correct handler on the client it must have an ID
+				// receive responses too.
+				err = l.conn.Call(
```

Comment:
Was this the only change from before? Because I'm like 99% sure I tried this too, but it didn't make a difference 😄

Reply:
The main diff is that this is A) Call and not notify, and B) does not happen while processing the request. This needs to be sent async.

Reply:
I guess it was the latter then... but how does the client know it's sent asynchronously, lol

Reply:
That doesn't sound very asynchronous to me, haha, but you're clearly correct 👍